File: livekit/services/redis_service.py
import os
import json
import redis
from dotenv import load_dotenv

load_dotenv()

r = redis.Redis(
    host=os.getenv("REDIS_HOST"),
    port=int(os.getenv("REDIS_PORT", 6379)),
    password=os.getenv("REDIS_PASS"),
    username=os.getenv("REDIS_USER"),
    decode_responses=True  # returns string instead of bytes
)

def publish_call_complete(call_id, status):
    payload = {
        "callId": call_id,
        "status": status
    }
    # Publish to same channel Node.js subscriber is listening on
    r.publish("call:complete", json.dumps(payload))
    print(f"📢 Published to Redis: {payload}")

File: backend/services/redisService.js
const dotenv = require('dotenv');
dotenv.config();

const Redis = require('ioredis');


const dids = new Map();

const redisSub = new Redis({
  host: process.env.REDIS_HOST,
  port: process.env.REDIS_PORT,
  password: process.env.REDIS_PASS,
  username: process.env.REDIS_USER,
  retryStrategy: (times) => Math.min(times * 50, 2000),
});

redisSub.on('connect', () => console.log('✅ Redis subscriber connected'));
redisSub.on('error', (err) => console.error('❌ Redis subscriber error:', err));

async function subscribe(channels, onMessage) {
  const channelList = Array.isArray(channels) ? channels : [channels];

  await redisSub.subscribe(...channelList);
  console.log(`📡 Subscribed to: ${channelList.join(', ')}`);

  redisSub.on('message', (channel, message) => {
    try {
      onMessage(channel, message);
    } catch (err) {
      console.error('Error in message callback:', err);
    }
  });
}


const waitForCallComplete = async (callId) => {
    return new Promise((resolve, reject) => {
        dids.set(callId, {
            complete: () => {
                resolve(true);
            }
        });
    });
}

const startSubscription = async () => {
    subscribe("call:complete", (channel, message) => {
        const { callId,status } = JSON.parse(message);
        console.log(`Call ${callId} status changed to ${status}`);

        const details = dids.get(callId);
        if (details){
            details.complete();
            dids.delete(callId);
        } else {
            console.log(`Call ${callId} not found`);
        }
    });
}
module.exports = { startSubscription, dids, waitForCallComplete };
